records/2021-07-16.scalability.database-bottleneck.md

```diff
@@ -0,0 +1,187 @@
+the autovacuum action that triggers a performance slowdown of our database:
+
+* **autovacuum_scale_factor** (default: 0.2): we could use this to delay the vacuum, (suggestion 0.3).
+* **cost_delay**: We could use this to delay by increasing the threshold.
+
+### Manual execution of the vacuum process on the `notifications` table
+
+An option is to keep the autovacuum enabled but trigger a manual vacuum prior
+to traffic time. This would better prepare the database for upcoming load.
+
+Don't run `vacuum FULL`: this will take on most database resources and prevent
+proper operation of the database.
+
+### Disable the autovacuum process on the `notifications` table
+
+To avoid the autovacuum to process the `notifications` table during the daily
+blast of notifications, we could disable the autovacuum to prevent it from
+triggering. This need to be done carefully, as a maximum limit of 4 billion
+transactions can occur in PostgreSQL. The autovacuum cleanup reset this
+transaction number. If this threshold is reached, our database will be severely
+affected.
+
+This is a risky move and certainly not a long/mid term solution.
+
+### Database connections pooling
+
+To maximise the distribution of database connection, we can have a
+pooler that would sit in-between the Notify stack and the AWS RDS
+Aurora Postgres instances.
+
+RDS Proxies would exists for this purpose. Apparently, we could create an
+endpoint each for both readers and the writer. This could co-exist nicely
+with our existing application logic which dispatch database sessions to
+either the readers or writer.
+
+### Tweak the readiness and liveness Kubernetes probe
+
+About these two probes: liveness probe will check if your pod is
+up abnd running; failure to do so will restart it. Readiness will check if
+your pod can receive traffic; failure to do so will detach the pod from the
+load balancer. Both probes will be continually checked throughout the
+pods' life.
+
+At the moment, the liveness probe targets the simple healthcheck that skips
+the database check while the readiness targets the healthcheck with the
+database check.
+
+Because our readiness probe reports that our pods are not ready anymore and
+hence, get detached from the load balancer, which **then cause notifications to
+be lost forever**, we could tweak it on the short term to accommodate our current
+situation and stop the bleeding.
+
+Normally, a delayed database should not be normal. We know in our current
+situation though that the database is still up and running. We'll receive paging
+alerts anyway if it goes down. We should definitely put back the database
+check once the design is fixed.
+
+### Notifications table sharding
+
+_To be completed_.
+
+## Decision
+
+Our current recommendations at the time of this writing:
+
+1. Stream the notifications save from the API to a Celery task.
+2. Insert the notifications as a batch once #1 is done.
```

Comment:
Are there plans to investigate the heavy queries issue on the DB when the celery task has additional work to do?

Reply:
No plan for now. It's worth checking, but not where the database spend most of its time. The AWS performance insights dashboard indicates that commits cost the most.

![image](https://user-images.githubusercontent.com/805567/126378299-58c7ef60-9307-472c-98cd-2bf319dc1fb5.png)